src/common/ipc/ipc.ts

```diff
@@ -0,0 +1,146 @@
+// Inter-process communications (main <-> renderer)
+// https://www.electronjs.org/docs/api/ipc-main
+// https://www.electronjs.org/docs/api/ipc-renderer
+
+import { ipcMain, ipcRenderer, webContents, remote } from "electron";
+import { toJS } from "mobx";
+import { EventEmitter } from "ws";
+import logger from "../../main/logger";
+import { ClusterFrameInfo, clusterFrameMap } from "../cluster-frames";
+
+const subFramesChannel = "ipc:get-sub-frames";
+
+export type HandlerEvent<EM extends EventEmitter> = Parameters<Parameters<EM["on"]>[1]>[0];
+export type EventListener<E extends EventEmitter, T extends any[]> = (event: HandlerEvent<E>, ...args: T) => any;
+
+export function handleRequest(channel: string, listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any) {
+  ipcMain.handle(channel, listener);
+}
+
+export async function requestMain(channel: string, ...args: any[]) {
+  return ipcRenderer.invoke(channel, ...args);
+}
+
+/**
+ * Adds a listener to `source` that waits for the first IPC message with the correct
+ * argument data is sent.
+ * @param channel The channel to be listened on
+ * @param listener The function for the channel to be called if the args of the correct type
+ * @param verifier The function to be called to verify that the args are the correct type
+ */
+export function onceCorrect<
+  EM extends EventEmitter,
+  T extends any[],
+  L extends (event: HandlerEvent<EM>, ...args: T) => any
+>(
+  source: EM,
+  channel: string | symbol,
+  listener: L,
+  verifier: (args: unknown[]) => args is T
+): void {
+  function handler(event: HandlerEvent<EM>, ...args: unknown[]): void {
+    if (verifier(args)) {
+      source.removeListener(channel, handler); // remove immediately
+
+      Promise.resolve(listener(event, ...args)) // might return a promise
+        .catch(error => logger.error("[IPC]: channel once handler threw error", { channel, error }));
+    } else {
+      logger.error("[IPC]: channel was sent to with invalid data", { channel, args });
+    }
+  }
+
+  source.on(channel, handler);
+}
+
+/**
+ * Adds a listener to `source` that checks to verify the arguments before calling the handler.
+ * @param channel The channel to be listened on
+ * @param listener The function for the channel to be called if the args of the correct type
+ * @param verifier The function to be called to verify that the args are the correct type
+ */
+export function onCorrect<
```

Comment:
Do this function is necessary to be added in current PR? I'm worried that such addition would make auto-update issue quite complex.

Reply:
How does this add undue complexity. I find the use of this function much less complex because it helps enforce typing through the IPC channel (and logs when they differ).

Reply:
`onCorrect` sounds a bit abstract.. maybe we can improve names in future? (naming is hard)

Reply:
Sounds good about improving the names in the future. For reference I chose this name because it is sort of equivalent to `ipc.on` but guarantees `correct` arguments.